Add docs on using namespaces at the node level for grouping nodes. (#4518)
kedro-org/kedro: docs/source/nodes_and_pipelines/namespaces.md

# Reuse pipelines and group nodes with namespaces

In this section, we introduce namespaces - a powerful tool for grouping and isolating nodes. Namespaces are useful in two key scenarios:

> - **Reusing a Kedro pipeline:** If you need to reuse a pipeline with some modifications of inputs, outputs or parameters, Kedro does not allow direct duplication because all nodes within a project must have unique names. Using namespaces helps resolve this issue by isolating identical pipelines while also enhancing visualization in Kedro-Viz.

Reviewer: We use UK spelling instead of US in the docs (actually Vale left a comment on that too, so it's worth not just ignoring that check) 

```suggestion
- **Reusing a Kedro pipeline:** If you need to reuse a pipeline with some modifications of inputs, outputs or parameters, Kedro does not allow direct duplication because all nodes within a project must have unique names. Using namespaces helps resolve this issue by isolating identical pipelines while also enhancing visualisation in Kedro-Viz.
```

After executing the pipeline with `kedro run`, the visualisation with `kedro viz run` after collapsing will look like this:

![namespaces collapsed all](../meta/images/namespaces_collapsed_all.png)

## Group nodes with namespaces

> You can [namespace your pipeline](#what-is-a-namespace) to group nodes within your pipeline, providing enhanced visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping.

Reviewer 2: > You can namespace your pipeline to group nodes within your pipeline

It's not entirely clear to me what's the difference between "group nodes" in this sentence and "group specific nodes" in the following sentence. I don't have a much better suggestion though. Maybe just

> You can namespace an entire pipeline to provide enhanced visualisation in Kedro-Viz...
> However, if instead you want to group specific nodes and not an entire pipeline, you can...

?

Author: I agree, modified

However, if you don’t want to namespace an entire pipeline but instead wish to group specific nodes from one or multiple pipelines under the same namespace, you can achieve this by assigning the `namespace` option in the node creation function. For example, if you want to group the first two nodes of the `data_processing` pipeline from [Spaceflights tutorial](../tutorial/add_another_pipeline.md#data-science-pipeline) into the same namespace, you can do:

```python
#src/project_name/pipelines/data_science/pipeline.py

from kedro.pipeline import Pipeline, node, pipeline
from .nodes import create_model_input_table, preprocess_companies, preprocess_shuttles

def create_pipeline(**kwargs) -> Pipeline:
    return pipeline(
        [
            node(
                func=preprocess_companies,
                inputs="companies",
                outputs="preprocessed_companies",
                name="preprocess_companies_node",

>                 namespace="preprocessing", # Assigning the node to the "preprocessing" namespace

Reviewer 2: What are the effects of adding a namespace to a node? Do you also have to namespace the corresponding catalog entries? If not, then nodes behave differently from pipelines in this regard and maybe we should clarify it.

Author: I clarified that after the example, but you're right - it makes sense to add an explicit clarification here as well. I've added it.

            ),
            node(
                func=preprocess_shuttles,
                inputs="shuttles",
                outputs="preprocessed_shuttles",
                name="preprocess_shuttles_node",
                namespace="preprocessing", # Assigning the node to the "preprocessing" namespace
            ),
            node(
                func=create_model_input_table,
                inputs=["preprocessed_shuttles", "preprocessed_companies", "reviews"],
                outputs="model_input_table",
                name="create_model_input_table_node",
            ),
        ]
    )
```

As you can see in the example above, we are using the `namespace` option only in the first two node creation functions. In this case, only the node names themselves will be prefixed with `preprocessing.`, while the inputs, outputs, and parameters remain unchanged.

Whether the namespace is assigned at the pipeline or node level, you can execute it separately using:
```bash
kedro run --namespace=<your_namespace>
```

Additionally, Kedro-Viz will allow you to expand and collapse your namespace for better visualisation.

> > **Caution:** While this functionality is useful for grouping nodes easily without additional complexities that may arise when creating a fully namespaced pipeline, be aware that it can create a namespace that is **non-executable**. This happens if you group, for example, the first and last nodes of a pipeline without including the intermediate nodes in the same namespace. In such cases, attempting to execute that namespace with `kedro run --namespace=<your_namespace>` will result in an error, and your visualisation in Kedro-Viz will be broken.

Reviewer 2: That sounds a bit scary and makes me think why would the user want to use `kedro run --namespace=node_group` instead of `kedro run --tags=node_group`.

Author: This functionality is quite similar - using a namespace at the node level simply adds a prefix to the node name. However, the key benefit I see is the enhanced visualisation in Kedro-Viz.

After executing the pipeline with `kedro run`, the visualisation with `kedro viz run` after collapsing will look like this:

![namespaces collapsed all](../meta/images/namespaces_collapsed_all.png)

## Group nodes with namespaces

> You can [namespace an entire pipeline](#what-is-a-namespace) to enhance visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping. But if you want to group specific nodes rather than an entire pipeline, you can do so by assigning the `namespace` option in the node creation function. In this case, only the node name will be prefixed with "namespace_name.", while inputs, outputs, and parameters will remain unchanged. Note that this behavior differs from namespacing at the pipeline level.

Reviewer 2: ```suggestion
You can [namespace an entire pipeline](#what-is-a-namespace) to enhance visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping. But if you want to group specific nodes rather than an entire pipeline, you can do so by assigning the `namespace` option in the node creation function. In this case, only the node name will be prefixed with `namespace_name.`, while inputs, outputs, and parameters will remain unchanged. Note that this behavior differs from namespacing at the pipeline level.
```

            ),
            node(
                func=preprocess_shuttles,
                inputs="shuttles",
                outputs="preprocessed_shuttles",
                name="preprocess_shuttles_node",
                namespace="preprocessing", # Assigning the node to the "preprocessing" namespace
            ),
            node(
                func=create_model_input_table,
                inputs=["preprocessed_shuttles", "preprocessed_companies", "reviews"],
                outputs="model_input_table",
                name="create_model_input_table_node",
            ),
        ]
    )
```

As you can see in the example above, we are using the `namespace` option only in the first two node creation functions. In this case, only the node names themselves will be prefixed with `preprocessing.`, while the inputs, outputs, and parameters remain unchanged.

Whether the namespace is assigned at the pipeline or node level, you can execute it separately using:
```bash
kedro run --namespace=<your_namespace>
```

Additionally, Kedro-Viz will allow you to expand and collapse your namespace for better visualisation.

> > **Caution:** While this functionality is useful for grouping nodes easily without additional complexities that may arise when creating a fully namespaced pipeline, be aware that it can create a namespace that is **non-executable**. This happens if you group, for example, the first and last nodes of a pipeline without including the intermediate nodes in the same namespace. In such cases, attempting to execute that namespace with `kedro run --namespace=<your_namespace>` will result in an error, and your visualisation in Kedro-Viz will be broken.

Reviewer 3: Can this part be in a warning notation: like this https://github.com/kedro-org/kedro/blob/7b4be6b53e70788f5af3c55d01b60280f3d9064f/docs/source/data/advanced_data_catalog_usage.md?plain=1#L5

Reviewer 3: There's some more sections in this page which are not modified in this PR which are formatted like this which could be formatted as ```{note} sections?

After executing the pipeline with `kedro run`, the visualisation with `kedro viz run` after collapsing will look like this:

![namespaces collapsed all](../meta/images/namespaces_collapsed_all.png)

## Group nodes with namespaces

> You can [namespace an entire pipeline](#what-is-a-namespace) to enhance visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping. But if you want to group specific nodes rather than an entire pipeline, you can do so by assigning the `namespace` option in the node creation function. In this case, only the node name will be prefixed with `namespace_name.`, while inputs, outputs, and parameters will remain unchanged. Note that this behavior differs from namespacing at the pipeline level.

Reviewer: ```suggestion
You can [namespace an entire pipeline](#what-is-a-namespace) to enhance visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping. But if you want to group specific nodes rather than an entire pipeline, you can do so by assigning the `namespace` argument in the node creation function. In this case, only the node name will be prefixed with `namespace_name.`, while inputs, outputs, and parameters will remain unchanged. Note that this behaviour differs from namespacing at the pipeline level.
```

> You can [namespace an entire pipeline](#what-is-a-namespace) to enhance visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping. But if you want to group specific nodes rather than an entire pipeline, you can do so by assigning the `namespace` option in the node creation function. In this case, only the node name will be prefixed with `namespace_name.`, while inputs, outputs, and parameters will remain unchanged. Note that this behavior differs from namespacing at the pipeline level.

Reviewer: I would maybe elaborate a bit on the last sentence and explain more clearly why this behaviour is different. Maybe we can leave it for another PR, but I noticed that in general we don't very clearly explain how you can map your inputs/outputs/parameters so you don't need to add namespaces in the catalog.

Author: There was a brief explanation at the top of the current docs page about how it works. I expanded that explanation to make it clearer and more explicit, and I added a link to it from the last sentence you mentioned.

            ),
            node(
                func=preprocess_shuttles,
                inputs="shuttles",
                outputs="preprocessed_shuttles",
                name="preprocess_shuttles_node",
                namespace="preprocessing", # Assigning the node to the "preprocessing" namespace
            ),
            node(
                func=create_model_input_table,
                inputs=["preprocessed_shuttles", "preprocessed_companies", "reviews"],
                outputs="model_input_table",
                name="create_model_input_table_node",
            ),
        ]
    )
```

As you can see in the example above, we are using the `namespace` option only in the first two node creation functions. In this case, only the node names themselves will be prefixed with `preprocessing.`, while the inputs, outputs, and parameters remain unchanged.

Whether the namespace is assigned at the pipeline or node level, you can execute it separately using:
```bash
kedro run --namespace=<your_namespace>
```

Additionally, Kedro-Viz will allow you to expand and collapse your namespace for better visualisation.

```{warning}

> While this functionality is useful for grouping nodes easily without additional complexities that may arise when creating a fully namespaced pipeline, be aware that it can create a namespace that is **non-executable**. This happens if you group, for example, the first and last nodes of a pipeline without including the intermediate nodes in the same namespace. In such cases, attempting to execute that namespace with `kedro run --namespace=<your_namespace>` will result in an error, and your visualisation in Kedro-Viz will be broken.

Reviewer: We should avoid saying things are "easy", because it might not be so easy or straightforward to a user reading this for the first time 🙂 

```suggestion
While this functionality is useful for grouping nodes without additional complexities that may arise when creating a fully namespaced pipeline, be aware that it can create a namespace that is **non-executable**. This happens if you group, for example, the first and last nodes of a pipeline without including the intermediate nodes in the same namespace. In such cases, attempting to execute that namespace with `kedro run --namespace=<your_namespace>` will result in an error, and your visualisation in Kedro-Viz will be broken.
```

After executing the pipeline with `kedro run`, the visualisation with `kedro viz run` after collapsing will look like this:

![namespaces collapsed all](../meta/images/namespaces_collapsed_all.png)

## Group nodes with namespaces

> You can [namespace an entire pipeline](#what-is-a-namespace) to enhance visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping. But if you want to group specific nodes rather than an entire pipeline, you can do so by assigning the `namespace` argument in the node creation function. In this case, only the node name will be prefixed with `namespace_name.`, while inputs, outputs, and parameters will remain unchanged. Note that this behavior differs from [namespacing at the pipeline level](#what-is-a-namespace).

Reviewer: Adding the link is a good idea 👍 Just one more change for UK spelling.

```suggestion
You can [namespace an entire pipeline](#what-is-a-namespace) to enhance visualisation in Kedro-Viz if your pipeline structure aligns with the desired node grouping. But if you want to group specific nodes rather than an entire pipeline, you can do so by assigning the `namespace` argument in the node creation function. In this case, only the node name will be prefixed with `namespace_name.`, while inputs, outputs, and parameters will remain unchanged. Note that this behaviour differs from [namespacing at the pipeline level](#what-is-a-namespace).
```
